### PR TITLE
Enable Estimated Stats on Query Plan for Instrumentation Monitoring

### PR DESCRIPTION
**Pull request by fgwang7w**
Test plan - (Please fill in how you tested your changes)
Enforce a testcase to verify the plan entirety with stats info using graphviz format and json format

```
== RELEASE NOTES ==

General Changes
* Add estimated stats on plan output for EXPLAIN format in GRAPHVIZ and JSON
```

**Before Change:**
```
                                                                                                                    Query Plan
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {
   "id" : "5",
   "name" : "Output",
   "identifier" : "[key1, value1]",
   "details" : "",
   "children" : [ {
     "id" : "80",
     "name" : "RemoteStreamingExchange",
     "identifier" : "[GATHER]",
     "details" : "",
     "children" : [ {
       "id" : "0",
       "name" : "TableScan",
       "identifier" : "[TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=default, tableName=test_grouped_join1, analyzePartitionValues=Optional.empty}', layout='Optional[default.test_grouped_join1{buckets=13}]'}]",
       "details" : "LAYOUT: default.test_grouped_join1{buckets=13}\nvalue1 := value1:varchar(79):1:REGULAR (1:37)\nkey1 := key1:bigint:0:REGULAR (1:37)\n",
       "children" : [ ],
       "remoteSources" : [ ]
     } ],
     "remoteSources" : [ ]
   } ],
   "remoteSources" : [ ]
 }
(1 row)

Query 20221212_231733_00000_2kdqx, FINISHED, 1 node
Splits: 1 total, 1 done (100.00%)
0:01 [0 rows, 0B] [0 rows/s, 0B/s]

```

**After Change:**
```
                                                                                                                    Query Plan
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {
   "id" : "5",
   "name" : "Output",
   "identifier" : "[key1, value1]",
   "details" : "",
   "children" : [ {
     "id" : "80",
     "name" : "RemoteStreamingExchange",
     "identifier" : "[GATHER]",
     "details" : "",
     "children" : [ {
       "id" : "0",
       "name" : "TableScan",
       "identifier" : "[TableHandle {connectorId='hive', connectorHandle='HiveTableHandle{schemaName=default, tableName=test_grouped_join1, analyzePartitionValues=Optional.empty}', layout='Optional[default.test_grouped_join1{buckets=13}]'}]",
       "details" : "LAYOUT: default.test_grouped_join1{buckets=13}\nvalue1 := value1:varchar(79):1:REGULAR (1:37)\nkey1 := key1:bigint:0:REGULAR (1:37)\n",
       "children" : [ ],
       "remoteSources" : [ ],
       "estimates" : [ {
         "outputRowCount" : 15000.0,
         "totalSize" : 1057364.0,
         "confident" : true,
         "variableStatistics" : {
           "value1<varchar(79)>" : {
             "lowValue" : "-Infinity",
             "highValue" : "Infinity",
             "nullsFraction" : 0.0,
             "averageRowSize" : 48.49093333333333,
             "distinctValuesCount" : 15000.0
           },
           "key1<bigint>" : {
             "lowValue" : 1.0,
             "highValue" : 60000.0,
             "nullsFraction" : 0.0,
             "averageRowSize" : "NaN",
             "distinctValuesCount" : 15000.0
           }
         }
       } ]
     } ],
     "remoteSources" : [ ],
     "estimates" : [ {
       "outputRowCount" : 15000.0,
       "totalSize" : 1057364.0,
       "confident" : true,
       "variableStatistics" : {
         "value1<varchar(79)>" : {
           "lowValue" : "-Infinity",
           "highValue" : "Infinity",
           "nullsFraction" : 0.0,
           "averageRowSize" : 48.49093333333333,
           "distinctValuesCount" : 15000.0
         },
         "key1<bigint>" : {
           "lowValue" : 1.0,
           "highValue" : 60000.0,
           "nullsFraction" : 0.0,
           "averageRowSize" : "NaN",
           "distinctValuesCount" : 15000.0
         }
       }
     } ]
   } ],
   "remoteSources" : [ ],
   "estimates" : [ {
     "outputRowCount" : 15000.0,
     "totalSize" : 1057364.0,
     "confident" : true,
     "variableStatistics" : {
       "value1<varchar(79)>" : {
         "lowValue" : "-Infinity",
         "highValue" : "Infinity",
         "nullsFraction" : 0.0,
         "averageRowSize" : 48.49093333333333,
         "distinctValuesCount" : 15000.0
       },
       "key1<bigint>" : {
         "lowValue" : 1.0,
         "highValue" : 60000.0,
         "nullsFraction" : 0.0,
         "averageRowSize" : "NaN",
         "distinctValuesCount" : 15000.0
       }
     }
   } ]
 }
(1 row)
```

### CODE REVIEW

**Comment by fgwang7w** [comment]
@aaneja could you please help give a first pass on this PR? Thanks!

**Comment by v-jizhang** [comment]
@bot kick off tests

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java [hunk 1]
```diff
@@ -185,6 +185,22 @@ public static String printDistributed(SubPlan plan, Session session, FunctionAnd
...
     }
+    public static String printDistributedFromFragments(List<PlanFragment> allFragments, Session session, FunctionAndTypeManager functionAndTypeManager)
+    {
+        PlanNodeIdGenerator idGenerator = new PlanNodeIdGenerator();
+        Map<PlanFragmentId, PlanFragment> fragmentsById = Maps.uniqueIndex(allFragments, PlanFragment::getId);
+
+        StringBuilder output = new StringBuilder();
+        output.append("digraph distributed_plan {\n");
+
+        for (PlanFragment planFragment : allFragments) {
```
Comment (aaneja) [inline presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java:196]: Should we sort `allFragments` by id so that the fragments are printed in order ? This won't impact rendered svg's, but will keep the text easier to read.

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java [hunk 2]
```diff
@@ -185,6 +185,22 @@ public static String printDistributed(SubPlan plan, Session session, FunctionAnd
 
         return output.toString();
     }
+    public static String printDistributedFromFragments(List<PlanFragment> allFragments, Session session, FunctionAndTypeManager functionAndTypeManager)
```
Comment (aaneja) [inline presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java:188]: Let's add a test in [TestGraphvizPrinter.java](https://github.com/prestodb/presto/blob/7d167d5fe7c1128ef95eb2fbee0b2ac3e80d6337/presto-main/src/test/java/com/facebook/presto/util/TestGraphvizPrinter.java) to exercise this as well

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java [hunk 3]
```diff
@@ -265,6 +265,21 @@ public void testOutputStats()
         assertEquals(1L, queryCompletedEvent.getStatistics().getOutputRows());
     }
 
+    @Test
+    public void testGraphvizQueryPlanOutput()
```
Comment (aaneja) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java:269]: Instead of having a new test, how about we add an assertion to [testNormalQuery](https://github.com/prestodb/presto/blob/ab6bdcb491a2578f0d432b74997f4df5d8db192e/presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java#L158) above and compare against a graphviz plan string that we get from a resource file ?

**Review by aaneja — CHANGES_REQUESTED** [review]
Some minor comments, but LGTM otherwise

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/JsonRenderer.java [hunk 4]
```diff
@@ -128,6 +133,12 @@ public List<String> getRemoteSources()
         {
             return remoteSources;
         }
+
+        @JsonProperty
+        public List<PlanNodeStatsEstimate> getEstimates()
```
[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/JsonRenderer.java:138]: yes because this is for deserialize to check for matching estimate which requires a serde for `VariableReferenceExpression`, I will a check in that commit.

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java (hunk 1 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java:196]: good point, in fact the sort should be done before it's passed in

File: presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java (hunk 3 above)
Comment (fgwang7w) [reply presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java:269]: I'd think to keep a separate unit test case because it needs to enable `print_stats_for_non_join_query` session property which I don't think it's related to the normalQuery testcase.

Comment (aaneja) [reply presto-tests/src/test/java/com/facebook/presto/execution/TestEventListener.java:269]: Cool - can you add a comment 'EXPLAIN will output graphviz plan with stats. To compare this against this the plan from the event listener graphviz output  `print_stats_for_non_join_query` needs to be set`  (Ignore my comment about using a resource file, I think its better this way - the graphviz node naming scheme can change over time)

**Comment by fgwang7w** [comment]
@bot kick off tests

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/JsonRenderer.java [hunk 5]
```diff
@@ -30,18 +37,29 @@
...
+    public JsonRenderer(FunctionAndTypeManager functionAndTypeManager)
+    {
+        JsonObjectMapperProvider provider = new JsonObjectMapperProvider();
+        provider.setJsonSerializers(ImmutableMap.of(VariableReferenceExpression.class, new Serialization.VariableReferenceExpressionSerializer()));
+        provider.setKeyDeserializers(ImmutableMap.of(VariableReferenceExpression.class, new Serialization.VariableReferenceExpressionDeserializer(functionAndTypeManager)));
+
+        JsonCodecFactory codecFactory = new JsonCodecFactory(provider, true);
+        this.codec = codecFactory.jsonCodec(JsonRenderedNode.class);
+        this.planMapCodec = codecFactory.mapJsonCodec(PlanFragmentId.class, JsonPlanFragment.class);
+    }
```
Comment (aaneja) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/JsonRenderer.java:52]: IMO we have sufficient complexity in this class now that it deserves it's own unit test demonstrating what the JSON output would be for a small distributed/logical plan

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 6]
```diff
@@ -193,16 +195,17 @@ public String toText(boolean verbose, int level)
         return new TextRenderer(verbose, level).render(representation);
     }
 
-    public String toJson()
+    public String toJson(FunctionAndTypeManager type)
     {
-        return new JsonRenderer().render(representation);
+        this.functionAndTypeManager = type;
+        return new JsonRenderer(type).render(representation);
```
Comment (aaneja) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:201]: Why can't we use the `functionAndTypeManager` that we instantiate with `PlanPrinter`

**Review by aaneja — COMMENTED** [review]
LGTM, minor comments

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java (hunk 6 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:201]: no you are right, this pass-in is not needed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/JsonRenderer.java (hunk 5 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/JsonRenderer.java:54]: make sense, add a new test coverage by `TestJsonRenderer`

File: presto-main/src/main/java/com/facebook/presto/sql/Serialization.java [hunk 7]
```diff
@@ -107,8 +108,13 @@ public void serialize(VariableReferenceExpression value, JsonGenerator jsonGener
     public static class VariableReferenceExpressionDeserializer
             extends KeyDeserializer
     {
-        private final TypeManager typeManager;
+        private TypeManager typeManager = createTestFunctionAndTypeManager();
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/Serialization.java:111]: Let's not make typeManager to be a "Test" one. Was there any reason you initialize it this way?

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/ExplainVerification.java [hunk 8]
```diff
@@ -45,7 +50,7 @@
         extends AbstractVerification<QueryBundle, ExplainMatchResult, String>
 {
     private static final ResultSetConverter<String> QUERY_PLAN_RESULT_SET_CONVERTER = resultSet -> Optional.of(resultSet.getString("Query Plan"));
-    private static final JsonCodec<JsonRenderedNode> PLAN_CODEC = jsonCodec(JsonRenderedNode.class);
+    private static JsonCodec<JsonRenderedNode> PLAN_CODEC;
```
Comment (yingsu00) [inline presto-verifier/src/main/java/com/facebook/presto/verifier/framework/ExplainVerification.java]: renaming it to planCodec should be done here in this commit, not the last one.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.common.type.VarcharType.VARCHAR;
+import static com.facebook.presto.metadata.AbstractMockMetadata.dummyMetadata;
+import static com.facebook.presto.metadata.FunctionAndTypeManager.createTestFunctionAndTypeManager;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.testng.Assert.assertEquals;
+
+public class TestJsonRenderer
+{
+    private static final FunctionAndTypeManager functionAndTypeManager = createTestFunctionAndTypeManager();
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:56]: Directly import and use `com.facebook.presto.testing.TestingEnvironment.FUNCTION_AND_TYPE_MANAGER`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    private static final ColumnHandle COLUMN_HANDLE = new TestingMetadata.TestingColumnHandle("column");
+    private static final PlanVariableAllocator variableAllocator = new PlanVariableAllocator();
+    private static final JsonRenderer jsonRenderer = new JsonRenderer(functionAndTypeManager);
+    private static final JsonObjectMapperProvider provider = new JsonObjectMapperProvider();
+    private static final TableHandle TABLE_HANDLE_WITH_LAYOUT = new TableHandle(
+            new ConnectorId("testConnector"),
+            new TestingMetadata.TestingTableHandle(),
+            TestingTransactionHandle.create(),
+            Optional.of(TestingHandle.INSTANCE));
+    private static JsonCodec<JsonRenderer.JsonRenderedNode> planCodec;
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:68]: final

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static org.testng.Assert.assertEquals;
+
+public class TestJsonRenderer
+{
+    private static final FunctionAndTypeManager functionAndTypeManager = createTestFunctionAndTypeManager();
+    private static final PlanBuilder PLAN_BUILDER = new PlanBuilder(TEST_SESSION, new PlanNodeIdAllocator(), dummyMetadata());
+    private static final VariableReferenceExpression COLUMN_VARIABLE = new VariableReferenceExpression(Optional.empty(), "column", VARCHAR);
+    private static final ColumnHandle COLUMN_HANDLE = new TestingMetadata.TestingColumnHandle("column");
+    private static final PlanVariableAllocator variableAllocator = new PlanVariableAllocator();
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:60]: variableAllocator, jsonRenderer should all be uppercased

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    private static final JsonRenderer jsonRenderer = new JsonRenderer(functionAndTypeManager);
+    private static final JsonObjectMapperProvider provider = new JsonObjectMapperProvider();
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:62]: The provider is only used once and doesn't have to be a class member

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+            new TestingMetadata.TestingTableHandle(),
+            TestingTransactionHandle.create(),
+            Optional.of(TestingHandle.INSTANCE));
+    private static JsonCodec<JsonRenderer.JsonRenderedNode> planCodec;
+
+    static {
+        provider.setKeyDeserializers(ImmutableMap.of(VariableReferenceExpression.class, new Serialization.VariableReferenceExpressionDeserializer(createTestFunctionAndTypeManager())));
+        JsonCodecFactory codecFactory = new JsonCodecFactory(provider, true);
+        planCodec = codecFactory.jsonCodec(JsonRenderer.JsonRenderedNode.class);
+    }
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:74]: nit: add empty line below

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    private PlanRepresentation getPlanRepresentation(PlanNode root)
+    {
+        return new PlanRepresentation(
+                root,
+                variableAllocator.getTypes(),
+                Optional.empty(),
+                Optional.empty());
+    }
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:82]: nit: add empty line below

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+                root.getClass().getName(),
+                root.getClass().getSimpleName(),
+                "",
+                root.getOutputVariables(),
+                Optional.empty(),
+                ImmutableList.of(),
+                ImmutableList.of(),
+                planNodeIds,
+                ImmutableList.of());
+    }
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:97]: nit: add empty line below

Comment (yingsu00) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:97]: All functions shall have an empty line below. You can use "Code -> Reformat Code" in IntelliJ to format the whole file, or all files in a folder

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+                ImmutableList.of(COLUMN_VARIABLE),
+                ImmutableMap.of(COLUMN_VARIABLE, COLUMN_HANDLE));
+        LimitNode limit = PLAN_BUILDER.limit(10, scan);
+        ProjectNode root = PLAN_BUILDER.project(new Assignments(map), limit);
+
+        return buildPlanResult(root);
+    }
+
+    @Test
+    public void testRoundtrip()
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:145]: Move the public methods to the top of the methods

**Review by yingsu00 — COMMENTED** [review]
@fgwang7w George, Could you paste some sample of the plan output for both before and after?

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java [hunk 17]
```diff
@@ -194,6 +197,8 @@ public VariableStatsEstimate getVariableStatistics(VariableReferenceExpression v
         return variableStatistics.getOrDefault(variable, VariableStatsEstimate.unknown());
     }
 
+    @JsonSerialize(keyUsing = Serialization.VariableReferenceExpressionSerializer.class)
+    @JsonDeserialize(keyAs = VariableReferenceExpression.class, keyUsing = Serialization.VariableReferenceExpressionDeserializer.class)
```
[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java:201]: I wonder if adding a blank line could cause any confusion, given that the serde annotation is associated to the method. If this is a general coding practice, I will add a blank line

File: presto-main/src/main/java/com/facebook/presto/sql/Serialization.java (hunk 7 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/Serialization.java:111]: this is used by test code path when jackson binder uses a default constructor that has no arg, I add a comment to clarify around the code block

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java (hunk 9 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:56]: noted for future reference, thanks

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java (hunk 12 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:62]: yes the provider is used only to generate the codec, will move it down to the static block

File: presto-verifier/src/main/java/com/facebook/presto/verifier/framework/ExplainVerification.java (hunk 8 above)
Comment (fgwang7w) [reply presto-verifier/src/main/java/com/facebook/presto/verifier/framework/ExplainVerification.java]: fixed

**Comment by fgwang7w** [comment]
@yingsu00 thank you for the review. I have updated the before/after change. All comments are addressed. Could you please help take a look again?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 18]
```diff
@@ -145,7 +147,7 @@
 public class PlanPrinter
 {
     private final PlanRepresentation representation;
-    private final FunctionAndTypeManager functionAndTypeManager;
+    private FunctionAndTypeManager functionAndTypeManager;
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:150]: Why removing the final keyword? It won't be reassigned anywhere else. 
Also, if the access modifiers are changed in any case, make sure they are grouped properly.

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java:201]: I don't remember how this happened, but I meant after the method definitions. You can disregard this one. Sorry for the confusion.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 19]
```diff
@@ -195,14 +197,14 @@ public String toText(boolean verbose, int level)
 
     public String toJson()
     {
-        return new JsonRenderer().render(representation);
+        return new JsonRenderer(functionAndTypeManager).render(representation);
     }
 
-    public static String jsonFragmentPlan(PlanNode root, Set<VariableReferenceExpression> variables, FunctionAndTypeManager functionAndTypeManager, Session session)
+    public static String jsonFragmentPlan(PlanNode root, Set<VariableReferenceExpression> variables, FunctionAndTypeManager functionAndTypeManager, StatsAndCosts estimatedStatsAndCosts, Session session)
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:203]: Can we reorder the arguments list and move estimatedStatsAndCosts to be after variables? The first 2 arguments and estimatedStatsAndCosts are all inputs/objects that need to be output, functionAndTypeManager and session are assisting objects.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 20]
```diff
@@ -303,18 +305,18 @@ public static String jsonLogicalPlan(
         return new PlanPrinter(plan, types, stageExecutionStrategy, functionAndTypeManager, estimatedStatsAndCosts, session, stats).toJson();
     }
 
-    public static String jsonDistributedPlan(StageInfo outputStageInfo)
+    public static String jsonDistributedPlan(FunctionAndTypeManager functionAndTypeManager, StageInfo outputStageInfo)
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:308]: Same as above, let's reorder the argument list.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 21]
```diff
@@ -303,18 +305,18 @@ public static String jsonLogicalPlan(
...
         List<PlanFragment> allFragments = getAllStages(Optional.of(outputStageInfo)).stream()
                 .map(StageInfo::getPlan)
                 .map(Optional::get)
                 .collect(toImmutableList());
-        return formatJsonFragmentList(allFragments);
+        return formatJsonFragmentList(functionAndTypeManager, allFragments);
     }
 
-    public static String jsonDistributedPlan(SubPlan plan)
+    public static String jsonDistributedPlan(FunctionAndTypeManager functionAndTypeManager, SubPlan plan)
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:317]: Same as above, let's reorder the argument list.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 22]
```diff
@@ -335,15 +337,15 @@ private String formatSourceLocation(Optional<SourceLocation> sourceLocation)
         return "";
     }
 
-    private static String formatJsonFragmentList(List<PlanFragment> fragments)
+    private static String formatJsonFragmentList(FunctionAndTypeManager functionAndTypeManager, List<PlanFragment> fragments)
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:340]: Same as above, let's reorder the argument list.

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java (hunk 2 above)
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java:188]: Let's be consistent with the ordering of the argument functionAndTypeManager and session. In PlanPrinter constructor, the ordering was reversed. Same for all other occurrences.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+    private static final VariableReferenceExpression COLUMN_VARIABLE = new VariableReferenceExpression(Optional.empty(), "column", VARCHAR);
+    private static final ColumnHandle COLUMN_HANDLE = new TestingMetadata.TestingColumnHandle("column");
+    private static final PlanVariableAllocator VARIABLE_ALLOCATOR = new PlanVariableAllocator();
+    private static final JsonRenderer JSON_RENDERER = new JsonRenderer(FUNCTION_AND_TYPE_MANAGER);
+    private static final TableHandle TABLE_HANDLE_WITH_LAYOUT = new TableHandle(
+            new ConnectorId("testConnector"),
+            new TestingMetadata.TestingTableHandle(),
+            TestingTransactionHandle.create(),
+            Optional.of(TestingHandle.INSTANCE));
+    private static final JsonCodec<JsonRenderer.JsonRenderedNode> planCodec;
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/planPrinter/TestJsonRenderer.java:65]: uppercase name

File: presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/cost/PlanNodeStatsEstimate.java:201]: thanks for clarification

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java (hunk 18 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:150]: no this variable should be marked as `final` and shouldn't be removed, fixed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java (hunk 19 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:203]: ok, reordered the parameters per suggestions

File: presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java [hunk 24]
```diff
@@ -183,6 +183,7 @@ private SubPlan buildFragment(PlanNode root, FragmentProperties properties, Plan
             schedulingOrder = scheduleOrder(root);
         }
 
+        StatsAndCosts stats = statsAndCosts.getForSubplan(root);
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java:186]: stats is only used once below. Inline this line into line 197

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 25]
```diff
@@ -2431,9 +2430,6 @@ private Optional<LocalDynamicFilter> createDynamicFilter(PhysicalOperation build
             if (node.getDynamicFilters().isEmpty()) {
                 return Optional.empty();
             }
-            if (buildSource.getPipelineExecutionStrategy() == GROUPED_EXECUTION) {
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:2434]: Why is this removed?

File: presto-main/src/test/java/com/facebook/presto/util/TestGraphvizPrinter.java [hunk 26]
```diff
@@ -117,6 +120,31 @@ public void testPrintDistributed()
         assertEquals(actualNestedSubPlan, expectedNestedSubPlan);
     }
 
+    @Test
+    public void testprintDistributedFromFragments()
```
Comment (yingsu00) [inline presto-main/src/test/java/com/facebook/presto/util/TestGraphvizPrinter.java:124]: uppercase P

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java (hunk 25 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:2434]: not sure why it's removed on the git, my local branch has this code piece, I will rebase to master and push the commit again

File: presto-main/src/main/java/com/facebook/presto/sql/Serialization.java (hunk 7 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/Serialization.java:113]: Unit tests from testcase `TestExplainVerification` would run into this code path when `TestingPrestoServer` boostraps `JsonCodecProvider`

**Comment by fgwang7w** [comment]
Thank you @yingsu00 for helping review the PR. All comments are addressed. could you please help give another pass? Thanks!

**Comment by fgwang7w** [comment]
@prestodb/committers ping!

**Review by aaneja — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/Serialization.java (hunk 7 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/Serialization.java:113]: @yingsu00 thanks for suggestion. I have refactored the code to remove the TestingFunctionAndTypeManager from mainline logic and add a mockup deserializer for testing purpose to enable test coverage.

File: presto-main/src/main/java/com/facebook/presto/sql/Serialization.java [hunk 27]
```diff
@@ -107,7 +107,15 @@ public void serialize(VariableReferenceExpression value, JsonGenerator jsonGener
     public static class VariableReferenceExpressionDeserializer
             extends KeyDeserializer
     {
-        private final TypeManager typeManager;
+        // when it is on the test code path, jackson deserializer instantiates the default constructor with no arg,
+        // the typeManager is initialized to use createTestFunctionAndTypeManager for test logic path only
+        private TypeManager typeManager; // = createTestFunctionAndTypeManager();
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/Serialization.java:112]: Comments not removed?

File: presto-main/src/main/java/com/facebook/presto/sql/Serialization.java [hunk 28]
```diff
@@ -107,7 +107,15 @@ public void serialize(VariableReferenceExpression value, JsonGenerator jsonGener
     public static class VariableReferenceExpressionDeserializer
             extends KeyDeserializer
     {
-        private final TypeManager typeManager;
+        // when it is on the test code path, jackson deserializer instantiates the default constructor with no arg,
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/Serialization.java:110]: Is this paragraph still valid?

**Review by yingsu00 — COMMENTED** [review]
@fgwang7w Could you please show me how you resolved the TypeManager initialization issue in presto-main/src/main/java/com/facebook/presto/sql/Serialization.java?

File: presto-main/src/main/java/com/facebook/presto/sql/Serialization.java (hunk 27 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/Serialization.java]: ok this code piece is cleaned up.

File: presto-main/src/main/java/com/facebook/presto/sql/Serialization.java (hunk 28 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/Serialization.java]: The fix to avoid using `createTestFunctionAndTypeManager` in `Serialization` main code path is to create a json mapper provider using the SerDe function for VariableReferenceExpression and use this custom json codec for serialization. This does not need any special no-args handling on the testing code path, e.g. TestExplainVerification, the `planCodec` is bind with `JsonObjectMapperProvider` that uses `VariableReferenceExpressionDeserializer`.

**Review by yingsu00 — APPROVED** [review]

**Comment by fgwang7w** [comment]
tests all green. Many thanks for reviewing and approving this PR. @yingsu00 @aaneja
